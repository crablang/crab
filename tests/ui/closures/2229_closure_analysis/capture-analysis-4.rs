// edition:2021

#![feature(crablangc_attrs)]

#[derive(Debug)]
struct Child {
    c: String,
    d: String,
}

#[derive(Debug)]
struct Parent {
    b: Child,
}

fn main() {
    let mut a = Parent { b: Child {c: String::new(), d: String::new()} };

    let c = #[crablangc_capture_analysis]
    //~^ ERROR: attributes on expressions are experimental
    //~| NOTE: see issue #15701 <https://github.com/crablang/crablang/issues/15701>
    || {
    //~^ First Pass analysis includes:
    //~| Min Capture analysis includes:
        let _x = a.b;
        //~^ NOTE: Capturing a[(0, 0)] -> ByValue
        //~| NOTE: Min Capture a[(0, 0)] -> ByValue
        println!("{:?}", a.b.c);
        //~^ NOTE: Capturing a[(0, 0),(0, 0)] -> ImmBorrow
    };
}
