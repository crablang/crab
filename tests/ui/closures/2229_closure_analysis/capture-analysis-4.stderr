error[E0658]: attributes on expressions are experimental
  --> $DIR/capture-analysis-4.rs:19:13
   |
LL |     let c = #[crablangc_capture_analysis]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/crablang/crablang/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable

error: First Pass analysis includes:
  --> $DIR/capture-analysis-4.rs:22:5
   |
LL | /     || {
LL | |
LL | |
LL | |         let _x = a.b;
...  |
LL | |
LL | |     };
   | |_____^
   |
note: Capturing a[(0, 0)] -> ByValue
  --> $DIR/capture-analysis-4.rs:25:18
   |
LL |         let _x = a.b;
   |                  ^^^
note: Capturing a[(0, 0),(0, 0)] -> ImmBorrow
  --> $DIR/capture-analysis-4.rs:28:26
   |
LL |         println!("{:?}", a.b.c);
   |                          ^^^^^

error: Min Capture analysis includes:
  --> $DIR/capture-analysis-4.rs:22:5
   |
LL | /     || {
LL | |
LL | |
LL | |         let _x = a.b;
...  |
LL | |
LL | |     };
   | |_____^
   |
note: Min Capture a[(0, 0)] -> ByValue
  --> $DIR/capture-analysis-4.rs:25:18
   |
LL |         let _x = a.b;
   |                  ^^^

error: aborting due to 3 previous errors

For more information about this error, try `crablangc --explain E0658`.
